### ISSUE TITLE
[DeepSpeed] Asking for feedback when training with zero2 with accelerate and diffusers

### ISSUE THREAD

**Issue by sayakpaul** [huggingface/accelerate#2787]
TL;DR: This is not really an issue reporting. This is more of asking for feedback if the changes I made to make `accelerate` and `deepspeed` work together. 

So, I have this training script: https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image.py. The following patch reflects the changes I made to it to make it work with DeepSpeed:

[ds.patch](https://github.com/huggingface/accelerate/files/15335196/ds.patch)

With these changes, I am able to successfully save and resume from checkpoints when using the following `accelerate` config (in a multi-GPU single-node setup):

```yaml
compute_environment: LOCAL_MACHINE
debug: false
deepspeed_config:
  gradient_accumulation_steps: 1
  offload_optimizer_device: none
  offload_param_device: none
  zero3_init_flag: false
  zero_stage: 2
distributed_type: DEEPSPEED
downcast_bf16: 'no'
gpu_ids: all
machine_rank: 0
main_training_function: main
mixed_precision: fp16
num_machines: 1
num_processes: 2
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```

## Example training commands

Start an initial run:

```bash
CUDA_VISIBLE_DEVICES=2,3 accelerate launch --config_file ds2.yaml train_text_to_image.py   --pretrained_model_name_or_path=$MODEL_NAME   --dataset_name=$DATASET_NAME   --mixed_precision="fp16"   --resolution=512 --center_crop --random_flip   --train_batch_size=1   --gradient_checkpointing   --max_train_steps=4   --learning_rate=1e-05   --checkpointing_steps=1 --output_dir=/raid/.cache/huggingface/finetuning-ds
```

Then resume from a checkpoint:

```bash
CUDA_VISIBLE_DEVICES=2,3 accelerate launch --config_file ds2.yaml train_text_to_image.py   --pretrained_model_name_or_path=$MODEL_NAME   --dataset_name=$DATASET_NAME   --mixed_precision="fp16"   --resolution=512 --center_crop --random_flip   --train_batch_size=1   --gradient_checkpointing   --max_train_steps=6   --learning_rate=1e-05 --resume_from_checkpoint="latest"  --checkpointing_steps=1 --output_dir=/raid/.cache/huggingface/finetuning-ds
```

Python dependencies to run the script can be found here: https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/requirements.txt. 

## `accelerate-cli env`

```bash
Copy-and-paste the text below in your GitHub issue

- `Accelerate` version: 0.30.1
- Platform: Linux-5.4.0-166-generic-x86_64-with-glibc2.31
- `accelerate` bash location: /home/sayak/.pyenv/versions/3.10.12/envs/diffusers/bin/accelerate
- Python version: 3.10.12
- Numpy version: 1.24.1
- PyTorch version (GPU?): 2.3.0+cu121 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- PyTorch MLU available: False
- System RAM: 503.54 GB
- GPU type: NVIDIA A100-SXM4-80GB
- `Accelerate` default config:
        Not found
```

LMK if you need more information here.

**Comment by sayakpaul** [comment]
Cc: @muellerzr @SunMarc

**Comment by SunMarc** [comment]
cc @pacman100

**Comment by sayakpaul** [comment]
Is there anyone looking into the DeepSpeed integration at the moment?

**Comment by SunMarc** [comment]
Sorry for the delay, there is a lot of backlog for deepspeed/FSDP issues since @pacman100 left but @muellerzr and I will be looking at these issues asap.

**Comment by sayakpaul** [comment]
Not stale

**Comment by sayakpaul** [comment]
Not stale

**Comment by muellerzr** [comment]
Hello hello! I think I know enough to answer yes/no with mild confidence now 🫡 Ty for your patience 🙇 

IIUC, the key areas of the diff are:

1. During model saving, you unwrap manually before calling `.save_pretrained`:
```diff
                 for i, model in enumerate(models):
-                    model.save_pretrained(os.path.join(output_dir, "unet"))
+                    if isinstance(unwrap_model(model), type(unwrap_model(unet))):
+                        unwrap_model(model).save_pretrained(os.path.join(output_dir, "unet"))
```
2. During loading you first `unwrap` the model and then call `register_to_config`
```diff
-                load_model = UNet2DConditionModel.from_pretrained(input_dir, subfolder="unet")
-                model.register_to_config(**load_model.config)
+                if isinstance(unwrap_model(model), type(unwrap_model(unet))):
+                    load_model = UNet2DConditionModel.from_pretrained(input_dir, subfolder="unet")
+                    model = unwrap_model(model)
+                    model.register_to_config(**load_model.config)
 
-                model.load_state_dict(load_model.state_dict())
-                del load_model
+                    model.load_state_dict(load_model.state_dict())
+                    del load_model
```

Re; 1: If your `save_pretrained` doesn't do this unwrapping, yes this is indeed the proper way to do so. This is exactly how transformers is setup:

```python
            if isinstance(self.accelerator.unwrap_model(self.model), supported_classes):
                self.accelerator.unwrap_model(self.model).save_pretrained(
                    output_dir, state_dict=state_dict, safe_serialization=self.args.save_safetensors
                )
```

Re; 2:

This also is correct, because we need to get access to the lowest level model to add it on. 

If you'd like to test to be certain that the wrapped model has your modifications, you can look in `model.module.module.xxx.mything` (for as many `.module` there are)

**Comment by sayakpaul** [comment]
Ah alright. Thanks for the confirmation! 

This thread will be helpful for folks who use `diffusers` to train with DeepSpeed, hopefully. Cc @bghira as well. I know you have workarounds but just wanted to make you aware.

**Comment by kopyl** [comment]
I have no idea what my CUDA_HOME env var shall be

**Comment by kopyl** [comment]
@sayakpaul installing nvcc with `apt install nvidia-cuda-toolkit` helped installing deepspeed

**Comment by kopyl** [comment]
@bghira thanks, but i think i'd rather wait for someone to fix the HF trainer than to figure out how yet another (hopefully working) library works.

**Comment by bghira** [comment]
the huggingface example training scripts are "thin", and do not contain all the necessary bits of logic to actually use DeepSpeed, and are not intended on being a complete example for eg. multigpu or ZeRO training.

**Comment by kopyl** [comment]
@bghira but simpletuner does not also have a complete guide on how to do a multi-gpu Flux Dreambooth training.

**Comment by bghira** [comment]
did ya look? 
https://github.com/bghira/SimpleTuner/blob/main/documentation/DEEPSPEED.md

**Comment by kopyl** [comment]
@bghira there is no word like "flux" or even "dreambooth" in that article.

**Comment by bghira** [comment]
jeez i didn't know i'd have to [hold your hand every step of the way](https://github.com/bghira/SimpleTuner/blob/main/documentation/quickstart/FLUX.md) maybe it's better you don't switch to ST :D

**Comment by kopyl** [comment]
@bghira sorry, i might be dumb, but i was unable to find how to launch the training. Only how to configure it.

**Comment by kopyl** [comment]
@bghira do you have a Docker container that has everything ready for a Dreambooth flux training including the complete guide on how to start it?
